File: python/cms.py
import datetime, json, requests

def user_count(posts):
    return len({post["userId"] for post in posts})

def summarise(post_count, user_count):
    return json.dumps({
        "posts": post_count,
        "users": user_count,
        "mean_posts_per_user": round(post_count / user_count)
    })

def store_cms_summary_report():
    print("Getting CMS data")
    response = requests.get("https://jsonplaceholder.typicode.com/posts")
    if response.status_code == 200:
        try:
            posts = json.loads(response.text)
            print("CMS data has %s records" % len(posts))
        except Exception as e:
            print('No valid JSON response from CMS, error %s' % type(e).__name__)
        today = datetime.date.today().isoformat()
        report_filename = "cms-" + today + ".json"
        with open(report_filename, "w") as summary_file:
            summary_file.write(summarise(len(posts), user_count(posts)))
            print("Wrote CMS report", flush=True)
    else:
        print('Error getting data from CMS, response code %s' % response.status_code)


if __name__ == '__main__':
    store_cms_summary_report()

File: java/src/main/java/CMS.java
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.*;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import static java.nio.charset.StandardCharsets.UTF_8;
import static java.time.format.DateTimeFormatter.ISO_DATE;

public class CMS {
    private static final String REPORT_DIRECTORY = System.getProperty("java.io.tmpdir");
    private static final String REPORT_NAME = String.format("cms-%s.json", LocalDateTime.now().format(ISO_DATE));
    private static final File REPORT_FILE = new File(REPORT_DIRECTORY, REPORT_NAME);

    public void storeCmsSummaryReport() {
        System.out.println("Getting CMS data");
        String postsText = "";
        try {
            URL postsUrl = new URL("https://jsonplaceholder.typicode.com/posts");
            try (InputStream is = postsUrl.openStream()) {
                postsText = new String(is.readAllBytes(), UTF_8);
            }
        } catch (IOException ex) {
            System.err.printf("No valid JSON response from CMS, error %s", ex.getMessage());
        }

        JSONArray posts = new JSONArray(postsText);
        System.out.printf("CMS data has %d records%n", posts.length());

        List<Integer> userIds = new ArrayList<>();
        for(int i = 0; i < posts.length(); i++) {
            JSONObject post = (JSONObject)posts.get(i);
            int userId = post.getInt("userId");
            if (!userIds.contains(userId)) {
                userIds.add(userId);
            }
        }

        try(BufferedWriter writer = new BufferedWriter(new FileWriter(REPORT_FILE))) {
            int postCount = posts.length();
            int userCount = userIds.size();
            String summary = String.format("{ \"posts\": %d, \"users\": %d, \"mean_posts_per_user\": %d }",
                    postCount, userCount, Math.round((float)postCount / (float)userCount));
            writer.write(summary);
        } catch (IOException e) {
            System.err.printf("Unable to write report %s: %s%n",
                    REPORT_FILE.getAbsolutePath(), e.getMessage());
        }
        System.out.println("Wrote CMS report");
    }

    public static void main(String[] args) {
        new CMS().storeCmsSummaryReport();
    }
}
